pkg/jobs/registry.go

```diff
@@ -1319,6 +1319,38 @@ func (r *Registry) cleanupOldJobsPage(
 	return !morePages, maxID, nil
 }
 
+// DeleteTerminalJobByID deletes the given job ID if it is in a
+// terminal state. If it is is in a non-terminal state, an error is
+// returned.
+func (r *Registry) DeleteTerminalJobByID(ctx context.Context, id jobspb.JobID) error {
+	return r.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		row, err := txn.QueryRow(ctx, "get-job-status", txn.KV(),
+			"SELECT status FROM system.jobs WHERE id = $1", id)
```

Comment:
You could eliminate this SELECT query by changing the first delete to something like:

```go
rowsAffected, err := txn.Exec(
    ctx, "delete-job", txn.KV(),
    "DELETE FROM system.jobs WHERE id = $1 AND status IN ($2, $3, $4)",
    id, StatusSucceeded, StatusCanceled, StatusFailed,
)
if err != nil { /* ... */ }
if rowsAffected == 0 {
  return errors.Newf("job %d has non-terminal status: %q", id, status)
}
```